app/controllers/api/v1/concerns/study_aware.rb

```diff
@@ -7,18 +7,23 @@ module StudyAware
         def set_study
           @study = Study.any_of({accession: params[:study_id]},{id: params[:study_id]}).first
           if @study.nil? || @study.queued_for_deletion?
-            head 404 and return
+            head 404
           end
         end
 
         ##
         # Permission checks
         ##
         def check_study_view_permission
-          if !@study.public? && !api_user_signed_in?
-            head 401
-          else
-            head 403 unless @study.public? || @study.can_view?(current_api_user)
+          unless @study.public?
+            if !api_user_signed_in? && params[:reviewerSession].present?
+              reviewer = @study.reviewer_access
+              head 401 if reviewer.nil? || reviewer.expired? || !reviewer.session_valid?(params[:reviewerSession])
```

Comment:
I have another question here: is `reviewerSession` _tied_ to a particular study? I.e. if I pass an otherwise "valid" (non-expired) session in a request for an unrelated study, will it "reject" the request? It might be helpful to also see how such a URL would look like (i.e. does it contain the study ID which is not "verified" simultaneously with the session).

Reply:
In short - yes, it will reject it.  A `reviewerSession` param is tied to the `Study` through the `ReviewerAccess` object, and then each `ReviewerAccess` has many `ReviewerAccessSession` objects.  The url looks like this for the ExploreController, for instance:

```
[:hostname]/single_cell/api/v1/studies/[:id]/explore?reviewerSession=[:session_key]
```

In this case, `:id` is most usually an `:accession` ('SCP1') which is used to load the study.  This happens via `:set_study` before `:check_study_view_permissions` even fires.  Once loaded, then if the `:reviewerSession` param is present, _and_ there is no authenticated user, then it will be used to validate view permission.  The order of operations for API calls is this:

```
before_action :set_current_api_user!
before_action :set_study
before_action :check_study_view_permission
```

Reply:
Sounds good! We'd still need to pentest this, of course, just to verify there're no access control issues (as we've done for other projects).